### ISSUE TITLE
System not  booting into usable state after systemd 253 upgrade

### ISSUE THREAD

**Issue by drzee99** [systemd/systemd#26613]
### systemd version the issue has been seen with

253

### Used distribution

Arch Linux

### Linux kernel version used

6.1.12.arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

System booting up and starting same service as before upgrade, going through the same targets

### Unexpected behaviour you saw

System not booting and stopping after reaching multi-user.target. Never reaching Login Prompts target. System in unusable state.

### Steps to reproduce the problem

Boot any Arch Linux AMI from https://wiki.archlinux.org/title/Arch_Linux_AMIs_for_Amazon_Web_Services (Arch Std Kernel) listed here on any AWS available hardware/instance type.

After boot complete:
sudo pacman-key --init
sudo pacman-key --populate
sudo pacman -Suy 

The last command will performance a full system update that also installs the latest Systemd version 253 - the AMI is build with 252.5 and is working fine.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Before updating to using 252.5 systemd we have the following targets and services:

Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Local Encrypted Volumes.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Local Integrity Protected Volumes.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Path Units.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Remote File Systems.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Slice Units.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Swaps.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Local Verity Protected Volumes.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Preparation for Local File Systems.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target Local File Systems.
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Reached target System Time Set.
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Reached target Host and Network Name Lookups.
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Reached target System Initialization.
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Reached target Timer Units.
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Reached target Socket Units.
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Reached target Basic System.
Feb 25 06:25:16 ip-172-31-50-96 systemd[1]: Reached target Preparation for Network.
Feb 25 06:25:16 ip-172-31-50-96 systemd[1]: Reached target Network.
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Reached target Cloud-config availability.
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Reached target Network is Online.
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Reached target Login Prompts.
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Reached target Multi-User System.
Feb 25 06:25:21 ip-172-31-50-96 systemd[1]: Reached target Cloud-init target.


Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Create List of Static Device Nodes...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module configfs...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module drm...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module fuse...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Load Kernel Modules...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Generate network units from Kernel command line...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Remount Root and Kernel File Systems...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Coldplug All udev Devices...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Load/Save Random Seed...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Apply Kernel Variables...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Create System Users...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Create Static Device Nodes in /dev...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Rebuild Dynamic Linker Cache...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Journal Service...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Rule-based Manager for Device Events and Files...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Flush Journal to Persistent Storage...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Create Volatile Files and Directories...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Security Auditing Service...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Rebuild Journal Catalog...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Network Name Resolution...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Network Time Synchronization...
Feb 25 06:25:08 ip-172-31-50-96 systemd[1]: Starting Update is Completed...
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Starting Record System Boot/Shutdown in UTMP...
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Starting Initial cloud-init job (pre-networking)...
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Starting D-Bus System Message Bus...
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Starting Name Service Cache Daemon...
Feb 25 06:25:09 ip-172-31-50-96 systemd[1]: Starting User Login Management...
Feb 25 06:25:16 ip-172-31-50-96 systemd[1]: Starting Network Configuration...
Feb 25 06:25:17 ip-172-31-50-96 systemd[1]: Starting Wait for Network to be Configured...
Feb 25 06:25:17 ip-172-31-50-96 systemd[1]: Starting Hostname Service...
Feb 25 06:25:19 ip-172-31-50-96 systemd[1]: Starting Initial cloud-init job (metadata service crawler)...
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Starting Apply the settings specified in cloud-config...
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Starting Permit User Sessions...
Feb 25 06:25:20 ip-172-31-50-96 systemd[1]: Starting Execute cloud user/final scripts...
Feb 25 06:25:44 ip-172-31-50-96 systemd[1]: Starting User Runtime Directory /run/user/1000...
Feb 25 06:25:44 ip-172-31-50-96 systemd[1]: Starting User Manager for UID 1000...

After update to 253 we get:

Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Local Encrypted Volumes.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Local Integrity Protected Volumes.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Path Units.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Slice Units.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Swaps.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Local Verity Protected Volumes.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Preparation for Network.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Preparation for Local File Systems.
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Reached target Local File Systems.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target System Time Set.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Network.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Host and Network Name Lookups.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target System Initialization.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Timer Units.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Socket Units.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Basic System.
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Reached target Multi-User System.

Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Create List of Static Device Nodes...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module configfs...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module dm_mod...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module fuse...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load Kernel Module loop...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load Kernel Modules...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Generate network units from Kernel command line...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Remount Root and Kernel File Systems...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Coldplug All udev Devices...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Load/Save OS Random Seed...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Apply Kernel Variables...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Create System Users...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Create Static Device Nodes in /dev...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Rebuild Dynamic Linker Cache...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Journal Service...
Feb 25 06:28:14 ip-172-31-50-96 systemd[1]: Starting Rule-based Manager for Device Events and Files...
Feb 25 06:28:15 ip-172-31-50-96 systemd[1]: Starting Flush Journal to Persistent Storage...
Feb 25 06:28:15 ip-172-31-50-96 systemd[1]: Starting Network Configuration...
Feb 25 06:28:15 ip-172-31-50-96 systemd[1]: Starting Create Volatile Files and Directories...
Feb 25 06:28:16 ip-172-31-50-96 systemd[1]: Starting Rebuild Journal Catalog...
Feb 25 06:28:16 ip-172-31-50-96 systemd[1]: Starting Network Name Resolution...
Feb 25 06:28:16 ip-172-31-50-96 systemd[1]: Starting Network Time Synchronization...
Feb 25 06:28:16 ip-172-31-50-96 systemd[1]: Starting Record System Boot/Shutdown in UTMP...
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Starting Update is Completed...
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Starting D-Bus System Message Bus...
Feb 25 06:28:17 ip-172-31-50-96 systemd[1]: Starting Hostname Service...
```

**Comment by drzee99** [comment]
Note, that the AMIs are build as standard Arch Linux Systems using standard Arch Linux tooling and processes (pacstrap) essentially following this guide: https://wiki.archlinux.org/title/installation_guide#Installation with some additional packages installed as default. 

Only standard arch linux packages available in the arch linux package repos are installed

The following packages are installed apart from what pacstrap (pulling the following packages: base grub mkinitcpio) pulls in: systemd-sysvcompat dosfstools e2fsprogs exfatprogs ntfs-3g xfsprogs man which lsof reflector rsync vi python3 audit irqbalance openssh haveged cloud-init cloud-utils aws-cli jq 

There is no separate boot partition and a basic grub/legacy-boot setup is used (same issue is seen using an UEFI boot partition setup).

SSH is configured with no password (only certificate) authentication and the network is configured to:

/etc/systemd/network/20.ethernet

[Match]
Name = en* eth*
[Network]
DHCP = yes
[DHCP]
UseMTU = yes
UseDNS = yes
UseDomains = yes

after initial booting cloud-init will create a new network under: /etc/systemd/network/10-cloud-init-ens5.network
[Match]
MACAddress=<obfuscated>
Name=ens5

[Network]
DHCP=ipv4

The following services are explicitly enabled:

systemctl enable systemd-timesyncd.service
systemctl enable nscd.service
systemctl enable auditd.service
systemctl enable haveged.service
systemctl enable irqbalance.service
systemctl enable sshd.service
systemctl enable cloud-init.service
systemctl enable cloud-config.service
systemctl enable cloud-final.service
systemctl enable systemd-networkd
systemctl enable systemd-resolved
systemctl enable sshd.service

I can see from the logs on Systemd 253 that the network is started and an IP address is received and time is synced with NTP. But after that nothing further happens. 

At this moment at loss what is causing this issue.

**Comment by poettering** [comment]
This is not actionable to us. Please talk to the Arch Linux folks first.

**Comment by drzee99** [comment]
Not sure why this is not actionable. From the failure pattern that I observe this is directly related to the recent SystemD release. 

For instance one of the targets that are not reached is a SystemD target file, so why is it not reached?

There have been no changes to the system the .service .target or .socket files. Everything worked before 253 now behavior is changed. 

Looking at the change log for systemd I can not see what change would have this of an impact? 

The only one that changed behavior was "systemd-networkd-wait-online" - but I don't see how this impacted as the network is started just fine, gets a link and an IP. 

I'm happy to further dig into this and look at this issue. But will need some guidance as where to look - I have tried to increase the debug level, but dont see an option for blanket increasing the debug level for all systemd releated actions (at least not one that results on more log captured in the journal.)

**Comment by mrc0mmand** [comment]
Please provide at least debug logs from the offending boot - i.e. by adding `systemd.log_level=debug systemd.log_target=console` to the kernel command line or tweaking `LogLevel=debug` and `LogTarget=console` in `/etc/systemd/system.conf`. `systemd.journald.forward_to_console=1` (or `ForwardToConsole=yes` in `/etc/systemd/journald.conf` might be helpful as well, if the previous debug output is not enough.

**Comment by drzee99** [comment]
Please find debug logs from a boot that succeeds (with Systemd 252) and one that fails (with Systemd 253). Aware that the two kernels are different versions, however that does not play a part (it was late an I was too lazy to redo both with the same 6.2 kernel). 

Apart from the update to of Systemd (and the kernel) nothing has changed between the two boots. 

What are we looking for?

[Systemd-252.5-boot.success.log](https://github.com/systemd/systemd/files/10866030/Systemd-252.5-boot.success.log)
[Systemd-253-boot.fail.log](https://github.com/systemd/systemd/files/10866031/Systemd-253-boot.fail.log)

**Comment by mrc0mmand** [comment]
The main difference I see is that the generated getty units are not pulled into the transaction:

252.5:
```
/usr/lib/systemd/system-generators/systemd-getty-generator succeeded.
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/container-getty@.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/serial-getty@.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty-pre.target
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/console-getty.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty.target
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty@.service
Unit serial-getty@ttyS0.service has alias serial-getty@.service.
Unit getty@tty1.service has alias getty@.service.
Unit getty@tty1.service has alias autovt@tty1.service.
getty@tty1.service: Installed new job getty@tty1.service/start as 89
getty.target: Installed new job getty.target/start as 88
system-serial\x2dgetty.slice: Installed new job system-serial\x2dgetty.slice/start as 94
serial-getty@ttyS0.service: Installed new job serial-getty@ttyS0.service/start as 93
system-getty.slice: Installed new job system-getty.slice/start as 90
system-getty.slice changed dead -> active
system-getty.slice: Job 90 system-getty.slice/start finished, result=done
...
```

253:
```
/usr/lib/systemd/system-generators/systemd-getty-generator succeeded.
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/container-getty@.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/serial-getty@.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty-pre.target
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/console-getty.service
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty.target
unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/getty@.service
Unit serial-getty@ttyS0.service has alias serial-getty@.service.
Unit getty@tty1.service has alias getty@.service.
Unit getty@tty1.service has alias autovt@tty1.service.
```

Not sure why, though (yet).

**Comment by eworm-de** [comment]
We have a downstream bug report about this (https://bugs.archlinux.org/task/77688). Looks like this is already fixed in git main. Any candidats for the fix?

**Comment by drzee99** [comment]
After doing the bisection it would appear that

5bc244aaa90211ccd8370535274c266cdff6a1cb is the first bad commit
commit 5bc244aaa90211ccd8370535274c266cdff6a1cb
Date: Sun Sep 25 09:18:41 2022 +0200

chase-symlinks: Add chase_symlinks_at()

Is one of the problematic commits that breaks things.

While everything works fine with 252.r60801.15b4b4ed06 and fails with 252.r60802.5bc244aaa9 I also observed that everything is fine with a much later release that I went through as part of the bisect, 252.r60846.a92401d558

I find odd as it does not look like the bisect centered around that one ... maybe I need to repeat the bisect?

**Comment by keszybz** [comment]
2877b14720?

I don't think we should do debugging of downstream issues in this bug tracker. We have dozens of downstream distributions, and each one generates hundreds of bugs per months which are specific to their version and setup and downstream modifications. This just can't scale.  I would expect distribution users report the issue to their distro first if they are using a distro-built package.

**Comment by eworm-de** [comment]
> [2877b14](https://github.com/systemd/systemd/commit/2877b14720928a214403870c68279ee3f8cc843f)?

That one is included in `v253`, no?

I suggested c677e13c35ff2d475199bcc9432892396485b6b8 and accc26a0e543c9f9e2e5adc1ad6c5a2b9d64670d in downstream bug... But no report, yet.

> I don't think we should do debugging of downstream issues in this bug tracker. We have dozens of downstream distributions, and each one generates hundreds of bugs per months which are specific to their version and setup and downstream modifications. This just can't scale to try to handle this in the upstream bugtracker. I would expect distribution users report the issue to their distro first if they are using a distro-built package.

Well, there is no change in downstream packaging that could explain this... And it looks to be fixed in git main.

So not a downstream issue IMHO... But I am happy to pinpoint this if possible.

**Comment by keszybz** [comment]
I looked at those two, and they don't specifically say they actually fix something. I wasn't sure if they are just "general cleanups" or "bugfix cleanups", so I skipped them in the backport.

@DaanDeMeyer ?

**Comment by keszybz** [comment]
> > [2877b14](https://github.com/systemd/systemd/commit/2877b14720928a214403870c68279ee3f8cc843f)?
> 
> That one is included in `v253`, no?

Nope. c677e13c35ff2d475199bcc9432892396485b6b8 == v253-rc2-27-gc677e13c35

**Comment by drzee99** [comment]
After spending some time doing a reverse bisect starting from 253.r62851.7a67afe331 (working) i found that:

accc26a0e543c9f9e2e5adc1ad6c5a2b9d64670d is the first fixed commit
commit accc26a0e543c9f9e2e5adc1ad6c5a2b9d64670d
Date: Sun Jan 8 18:19:11 2023 +0100

chase-symlinks: Always open a dirfd to the root directory

Instead of special casing "/", let's just always open a dirfd to the
root directory.

253.r62382.c677e13c35 is NOT working and 253.r62383.accc26a0e5 is working.

So the conclusion must be that the change introduced in 252.r60802.5bc244aaa9 is what breaks things and its fixed again with 253.r62383.accc26a0e5 ...

**Comment by eworm-de** [comment]
Mystery solved, commits found, cherry-picked for `v253.1` and pushed in `systemd` package `253.1-1`.
I guess we can close this now?

**Comment by drzee99** [comment]
Thanks for the help.

I can confirm that running the systemd package from the Testing Repo (253.1-1) that everything works.
